src/GafferOSL/ShadingEngine.cpp

```diff
@@ -1297,5 +1297,11 @@ bool ShadingEngine::needsAttribute( const std::string &scope, const std::string
 		return true;
 	}
 
+	// globals which are always needed
+	if ( name == "P" || name == "uv" || name == "u" || name == "v" || name == "N" )
```

Comment:
I really want to limit the amount of unnecessary resampling and de-indexing we do - it's a major overhead. Can't we use the `globals_needed` query we're already using in `ShadingEngine::queryContextVariablesAndAttributesNeeded()` to make this more accurate?

Reply:
Is something like 3815e9c better?

Reply:
Yes. But I don't think the attribute/global distinction is relevant in terms of the ShadingEngine interface, because everything is passed in via that one container of points to be shaded. As a user of the class there is no distinction at all. I know `needsAttribute()` takes both `scope` and `name`, but actually that's misleading because `scope` is unused from the point of view of the client of the ShadingEngine. Likewise, the combined "uv" vs separate "u" and "v" consideration could be totally encapsulated within the ShadingEngine, and so could the requirement for "P". OSLObject just literally just need to call `ShadingEngine::needsAttribute( primVarName )` and nothing else.

As a separate thing, I think it'd be good to remove the requirement for "P" by passing in a point count on its own, as it could reduce overhead in a lot of OSLImage cases. But that can wait. Let's just get it working so that we have a single `ShadingEngine::needsAttribute( name )` call that encapsulates everything, hiding any distinction between globals and attributes...

Reply:
> Yes. But I don't think the attribute/global distinction is relevant in terms of the ShadingEngine interface, because everything is passed in via that one container of points to be shaded. As a user of the class there is no distinction at all. I know needsAttribute() takes both scope and name, but actually that's misleading because scope is unused from the point of view of the client of the ShadingEngine. Likewise, the combined "uv" vs separate "u" and "v" consideration could be totally encapsulated within the ShadingEngine, and so could the requirement for "P". OSLObject just literally just need to call ShadingEngine::needsAttribute( primVarName ) and nothing else.

right o

> As a separate thing, I think it'd be good to remove the requirement for "P" by passing in a point count on its own,

Agreed

Reply:
How about 78cff01 ?

Reply:
Looks good - thanks Don!